libs/rtemodel/src/RteGenerator.cpp

```diff
@@ -145,12 +145,18 @@ string RteGenerator::GetExecutable(RteTarget* target, const std::string& hostTyp
 vector<pair<string, string> > RteGenerator::GetExpandedArguments(RteTarget* target, const string& hostType) const
 {
   vector<pair<string, string> > args;
+  string packId = GetPackage()->GetPackYamlID();
+  string generatorId = GetGeneratorName();
+
   RteItem* argsItem = GetArgumentsItem("exe");
   if (argsItem) {
     for (auto arg : argsItem->GetChildren()) {
       if (arg->GetTag() != "argument" || !arg->MatchesHost(hostType))
         continue;
-      args.push_back({arg->GetAttribute("switch"), ExpandString(arg->GetText())});
+      string res = arg->GetText();
+      RteUtils::ReplaceAll(res, "$pack_id", packId);
+      RteUtils::ReplaceAll(res, "$generator_id", generatorId);
```

Comment:
It seems `$pack_id` and `$generator_id` are deviating from the standard way of declaring/expanding/using variables such as `$P #P $S $D $B $G`:
https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/pdsc_generators_pg.html#element_generators
Concerning implementation look at [`RteCallback::ExpandString`](https://github.com/Open-CMSIS-Pack/devtools/blob/main/libs/rtemodel/src/RteCallback.cpp#L67).

@slhultgren We have discussed about the use of `[Gvendor][::Gtool][@Gversion]` as generator id. As long as the vendor guarantees the `Gtool` uniqueness, wouldn't it be sufficient?
https://github.com/Open-CMSIS-Pack/devtools/issues/434#issuecomment-1337608400

Reply:
Hi @brondani,

I'm not sure to catch the point about Gvendor/Gtool/Gversion, since my problem currently with these are:
1. I don't understand what/how they are used
1.1. Gtool is defined as "Plain text name", this sounds almost like a human-readable description, not something to rely on functionally
1.2. There is no implementation yet supporting this AFAIK
2. They are still not guaranteed to be unique globally, so we would still require the pack id to make it fully unique
3. If Gvendor must follow the pack vendor, not sure why it is even possible to define it for the generator

Technically I guess Gvendor/Gtool/Gversion could work, as long as:
1. Open-CMSIS-Pack project guarantees that Only a single vendor can provide generators for a Gvendor
2. The Vendor guarantees that it does not release multiple conflicting generators using the same Gtool

but honestly I don't think any of these two points are feasible in practice (not worth the effort; The alternative of just allowing to specify the pack id is a lot simpler to maintain and double check on an organizational level. Each pack is more self-contained).

Currently my view on the Gvendor/Gtool/Gversion is that they provide some abstraction for calling generators (potentially from other packs??) but I don't think this level of abstraction is necessary or maybe not even useful?
Most likely when defining a component using a generator, we will know/expect some kind of behavior of the generator that we call, so most likely we just want to be able to call a generator from (at most) another pack (that we already know).

Reply:
@slhultgren 
>1.1. Gtool is defined as "Plain text name", this sounds almost like a human-readable description, not something to rely on functionally

Not really different from other identifiers such as the pack name itself.

> 1.2. There is no implementation yet supporting this AFAIK

Right, if we want to benefit from the spec we should start implementing it. I thought we agreed [here](https://github.com/Open-CMSIS-Pack/devtools/issues/434#issuecomment-1337608400) this was the best approach to follow?

> 2.	They are still not guaranteed to be unique globally, so we would still require the pack id to make it fully unique

`Gvendor` should be unique. Each vendor would be responsible for their own `Gtool` identifiers.

> 3.	If Gvendor must follow the pack vendor, not sure why it is even possible to define it for the generator

I guess for the same reason it’s possible to specify `Cvendor` for components: if omitted it can be inherited from pack vendor, but it does not have to.

If you now strongly believe the `from-pack` is a better approach, I would recommend to update the specification.